tests/devices/femto/test_femto.py

```diff
@@ -0,0 +1,101 @@
+import pytest
+from mock import Mock, create_autospec
+
+from tickit.core.device import DeviceUpdate
+from tickit.core.typedefs import SimTime
+from tickit.devices.femto.femto import CurrentDevice, Femto, FemtoAdapter
+
+
+@pytest.fixture
+def femto() -> Femto:
+    return Femto()
+
+
+def test_femto_constructor(femto: Femto):
+    pass
+
+
+def test_set_and_get_gain(femto: Femto):
+    femto.set_gain(3.0)
+    assert femto.get_gain() == 3.0
+
+
+def test_set_and_get_current(femto: Femto):
+    femto.set_current(3.0)
+    assert femto.get_current() == 3.0 * femto.get_gain()
+
+
+def test_femto_update(femto: Femto):
+    device_input = {"input": 3.0}
+    time = SimTime(0)
+    update: DeviceUpdate = femto.update(time, device_input)
+    assert update.outputs["current"] == 3.0 * femto.get_gain()
+
+
+# # # # # # # # # # CurrentDevice # # # # # # # # # #
+
+
+@pytest.fixture
+def current_device() -> CurrentDevice:
+    return CurrentDevice()
+
+
+def test_current_device_constructor(current_device: CurrentDevice):
+    pass
+
+
+def test_current_device_update(current_device: CurrentDevice):
+    time = SimTime(0)
+    device_input = {"bleep": "bloop"}
+    update: DeviceUpdate = current_device.update(time, device_input)
+    assert 0.1 <= update.outputs["output"] <= 200.1
+
+
+# # # # # # # # # # FemtoAdapter # # # # # # # # # #
+
+
+@pytest.fixture
+def mock_femto() -> Mock:
+    return create_autospec(Femto, instance=True)
+
+
+@pytest.fixture
+def raise_interrupt() -> Mock:
+    async def raise_interrupt():
+        return False
+
+    return Mock(raise_interrupt)
+
+
+@pytest.fixture
+def femto_adapter(mock_femto: Mock, raise_interrupt) -> FemtoAdapter:
+    return FemtoAdapter(mock_femto, raise_interrupt)
+
+
+def test_femto_adapter_constructor(femto_adapter: FemtoAdapter):
+    pass
+
+
+@pytest.mark.asyncio
+async def test_run_forever(femto_adapter: FemtoAdapter):
+    femto_adapter.build_ioc = Mock(femto_adapter.build_ioc)
+    await femto_adapter.run_forever()
+    femto_adapter.build_ioc.assert_called_once()
+
+
+@pytest.mark.asyncio
+async def test_femto_adapter_callback(femto_adapter: FemtoAdapter):
+    await femto_adapter.callback(2.0)
+    femto_adapter._device.set_gain.assert_called_with(2.0)
+    femto_adapter.raise_interrupt.assert_called()
+
+
+def test_on_db_load(femto_adapter: FemtoAdapter):
+    femto_adapter.on_db_load()
+
+    input_record = femto_adapter.input_record
+    current_record = femto_adapter.current_record
+    interrupt_records = femto_adapter.interrupt_records
+
+    assert interrupt_records[input_record] == femto_adapter._device.get_gain
```

Comment:
This assertion is failing. Is this because of the test or the implementation?

Reply:
Investigating...

Reply:
This test fails when run by `pipenv run tests` but passes when run by `pytest tests/devices/femto/test_femto.py`...

Reply:
So it also fails when running all the tests using pytest. Which is quite unfortunate. I suppose that means the tests are interacting somehow...?

Reply:
My guess is that it's something to do with the order in which things are loaded

Reply:
For the record, the issue is the mocking of the softioc library for the test_epics_adapter.py file.

Reply:
The issue is that since input record and current record are accessing methods of the same `Mock` object, their hashes are identical. Hence the value of input_record gets overwritten by the value of current record. Is there a way to return different hashes for different methods of a `Mock`?

Reply:
Are we wanting to mock these records? I had presumed not since it's only a side effect of the patch in `test_epics_adapter`.
Regardless, if you create a mock you can assign the `__hash__.return_value`; e.g.
```python
input_record = Mock()
input_record.__hash__.return_value = 42
```

Reply:
The above doesn't work since the hash will still be the same for both `input_record` and `current_record`, causing the overwriting bug.

Reply:
I've been trying the following snippet as an alternative to mocking the module:
```
mock_builder = patch("softioc.builder", autospec=True)
    mock_softioc = patch("softioc.softioc", autospec=True)
    mock_asyncio = patch("softioc.asyncio_dispatcher", autospec=True)
    mock_f = patch("builtins.open", mock_open(read_data=data))
    mock_unlink = patch("os.unlink", autospec=True)

    with mock_softioc, mock_builder, mock_softioc, mock_asyncio, mock_f, mock_unlink as unlink:
        epics_adapter.build_ioc()
        unlink_args = unlink.call_args.args
```
However pytest silently crashes out when running this test. 

:joy_cat:

Reply:
I think I'll let this test skip for now. It's a bit unreliable and can cause the whole test suite to break. We'll revisit this after an Epics Adapter refactoring.

Reply:
This femto test now works, the clashing test `test_epics_adapter_build_ioc_method` has been skipped for now due to it's tendency to crash the whole test suite.